### PR TITLE
document visibility feature and usage in sample Sugar uncon code

### PR DESCRIPTION
**Pull request by jzhu-sugar**

### CODE REVIEW

File: custom_visibility/src/custom/data/visibility/FilterOpportunities.php [hunk 1]
```diff
@@ -65,20 +65,30 @@ public function __construct(SugarBean $bean, $params = null)
     }
 
     /**
-     * {@inheritdoc}
+     * orvewrite parent method, SugarVisibility::addVisibilityWhere
```
Comment (lschaefer-sugarcrm) [inline custom_visibility/src/custom/data/visibility/FilterOpportunities.php]: Typo: should be Overwrite

File: custom_visibility/src/custom/data/visibility/FilterOpportunities.php [hunk 2]
```diff
@@ -65,20 +65,30 @@ public function __construct(SugarBean $bean, $params = null)
...
+     *
+     * add visibility, i.e. condition, clauses to the WHERE part of the query
+     *
+     * this demo function adds new condition to where clause with sales_stage is NOT IN the list of $this->filterSalesStages
```
Comment (lschaefer-sugarcrm) [inline custom_visibility/src/custom/data/visibility/FilterOpportunities.php]: I'd change *with* to *when*
"where clause *when* sales_stage is NOT IN..."

File: custom_visibility/src/custom/data/visibility/FilterOpportunities.php [hunk 3]
```diff
@@ -65,20 +65,30 @@ public function __construct(SugarBean $bean, $params = null)
...
+     *
+     * @param string $query
+     * @return string
```
Comment (lschaefer-sugarcrm) [inline custom_visibility/src/custom/data/visibility/FilterOpportunities.php]: Can you provide more info on what is being returned in this string?

File: custom_visibility/src/custom/data/visibility/FilterOpportunities.php [hunk 4]
```diff
@@ -89,11 +99,20 @@ public function addVisibilityWhere(&$query)
     }
 
     /**
-     * {@inheritdoc}
+     * orvewrite parent method, SugarVisibility::addVisibilityWhereQuery
```
Comment (lschaefer-sugarcrm) [inline custom_visibility/src/custom/data/visibility/FilterOpportunities.php]: Typo: should be Overwrite

File: custom_visibility/src/custom/data/visibility/FilterOpportunities.php [hunk 5]
```diff
@@ -89,11 +99,20 @@ public function addVisibilityWhere(&$query)
...
+     *
+     * add visibility, i.e. condition, clauses to the WHERE part of SugarQuery
+     *
+     * @param SugarQuery $query
+     * @param array $options
+     * @return SugarQuery
```
Comment (lschaefer-sugarcrm) [inline custom_visibility/src/custom/data/visibility/FilterOpportunities.php]: I'd add a few words to the return statement.  Perhaps something like "The updated query"

File: custom_visibility/src/custom/data/visibility/FilterOpportunities.php [hunk 6]
```diff
@@ -89,11 +99,20 @@ public function addVisibilityWhere(&$query)
...
      */
     public function addVisibilityWhereQuery(SugarQuery $sugarQuery, $options = array()) {
+        // create where clause
         $where = null;
         $this->addVisibilityWhere($where, $options);
+
+        // add the where clause to $sugarQuery's where part
         if (!empty($where)) {
```
Comment (lschaefer-sugarcrm) [inline custom_visibility/src/custom/data/visibility/FilterOpportunities.php]: What happens if the $where query is empty?

File: custom_visibility/src/custom/data/visibility/FilterOpportunities.php [hunk 7]
```diff
@@ -168,23 +187,42 @@ protected function getTableAlias()
     }
 
     /**
-     * {@inheritdoc}
+     * implementation of ElasticStrategyInterface interface
+     *
+     * to create ealstic analyzer for the fields introduced here,
```
Comment (lschaefer-sugarcrm) [inline custom_visibility/src/custom/data/visibility/FilterOpportunities.php]: Typo: should be elastic

Comment (lschaefer-sugarcrm) [inline custom_visibility/src/custom/data/visibility/FilterOpportunities.php]: I don't know what an elastic analyzer is.  Is that common knowledge?  Could we add a link to a reference or some words of explanation?

File: custom_visibility/src/custom/data/visibility/FilterOpportunities.php [hunk 8]
```diff
@@ -168,23 +187,42 @@ protected function getTableAlias()
...
+     * leave the function empty if you don't need special analyzer
```
Comment (lschaefer-sugarcrm) [inline custom_visibility/src/custom/data/visibility/FilterOpportunities.php]: lines 192 and 193 don't feel like a complete thought.  If they are separate thoughts, we should end 192 with a period.  I'm wondering if the word "to" on line 192 shouldn't be there?

File: custom_visibility/src/custom/data/visibility/FilterOpportunities.php (hunk 7 above)
Comment (lschaefer-sugarcrm) [inline custom_visibility/src/custom/data/visibility/FilterOpportunities.php]: What fields?  I don't see any in the params

File: custom_visibility/src/custom/data/visibility/FilterOpportunities.php [hunk 9]
```diff
@@ -168,23 +187,42 @@ protected function getTableAlias()
...
     public function elasticBuildAnalysis(AnalysisBuilder $analysisBuilder, Visibility $provider)
     {
         // no special analyzers needed
     }
 
     /**
-     * {@inheritdoc}
+     * implementation of ElasticStrategyInterface interface
+     *
+     * Build Elasticsearch mapping, it defines how a document's fields are stored and indexed in Elasitc engine
```
Comment (lschaefer-sugarcrm) [inline custom_visibility/src/custom/data/visibility/FilterOpportunities.php]: Perhaps change "it" to "which"

File: custom_visibility/src/custom/data/visibility/FilterOpportunities.php [hunk 10]
```diff
@@ -168,23 +187,42 @@ protected function getTableAlias()
...
     public function elasticBuildMapping(Mapping $mapping, Visibility $provider)
     {
         $mapping->addNotAnalyzedField('visibility_sales_stage');
     }
 
     /**
-     * {@inheritdoc}
+     * implementation of ElasticStrategyInterface interface
+     *
+     * Process document before its being indexed,
```
Comment (lschaefer-sugarcrm) [inline custom_visibility/src/custom/data/visibility/FilterOpportunities.php]: "its being indexed" 
I'd reword to
"it has been indexed"

File: custom_visibility/src/custom/data/visibility/FilterOpportunities.php [hunk 11]
```diff
@@ -194,7 +232,15 @@ public function elasticProcessDocumentPreIndex(Document $document, \SugarBean $b
     }
 
     /**
-     * {@inheritdoc}
+     * implementation of ElasticStrategyInterface interface
+     *
+     * Bean index fields to be indexed
```
Comment (lschaefer-sugarcrm) [inline custom_visibility/src/custom/data/visibility/FilterOpportunities.php]: I think this would be more clear if it were a verb phrase.  As written, it sounds like a param description.

File: custom_visibility/src/custom/data/visibility/FilterOpportunities.php [hunk 12]
```diff
@@ -194,7 +232,15 @@ public function elasticProcessDocumentPreIndex(Document $document, \SugarBean $b
...
+     *
+     * this demo function adds field, 'sales_stage', to be indexed
+     *
+     * @param string $module
+     * @param Visibility $provider
+     * @return array
```
Comment (lschaefer-sugarcrm) [inline custom_visibility/src/custom/data/visibility/FilterOpportunities.php]: It would be nice to have a few words describing what the params and return array are

File: custom_visibility/src/custom/data/visibility/FilterOpportunities.php [hunk 13]
```diff
@@ -203,7 +249,15 @@ public function elasticGetBeanIndexFields($module, Visibility $provider)
...
-     * {@inheritdoc}
+     * implementation of ElasticStrategyInterface interface
+     *
+     * Add visibility filters
+     *
+     * this demo function adds elastic filter with condition 'filter_sales_stages' NOT IN $this->filterSalesStages
+     *
+     * @param User $user
+     * @param \Elastica\Filter\Bool $filter
+     * @param Visibility $provider
```
Comment (lschaefer-sugarcrm) [inline custom_visibility/src/custom/data/visibility/FilterOpportunities.php]: Would be nice to have a few words describing these params

File: custom_visibility/src/custom/src/Elasticsearch/Provider/Visibility/Filter/OpportunitySalesStagesFilter.php [hunk 14]
```diff
@@ -22,15 +22,25 @@ class OpportunitySalesStagesFilter implements FilterInterface
     protected $provider;
 
     /**
-     * {@inheritdoc}
+     * implentation of FilterInterface::setProvider
```
Comment (lschaefer-sugarcrm) [inline custom_visibility/src/custom/src/Elasticsearch/Provider/Visibility/Filter/OpportunitySalesStagesFilter.php]: Typo: Implementation

File: custom_visibility_7_10/src/custom/data/visibility/FilterOpportunities.php [hunk 15]
```diff
@@ -65,20 +65,30 @@ public function __construct(SugarBean $bean, $params = null)
     }
 
     /**
-     * {@inheritdoc}
+     * orvewrite parent method, SugarVisibility::addVisibilityWhere
```
Comment (lschaefer-sugarcrm) [inline custom_visibility_7_10/src/custom/data/visibility/FilterOpportunities.php]: Typo: overwrite

File: custom_visibility_7_10/src/custom/data/visibility/FilterOpportunities.php [hunk 16]
```diff
@@ -65,20 +65,30 @@ public function __construct(SugarBean $bean, $params = null)
...
+     *
+     * add visibility, i.e. condition, clauses to the WHERE part of the query
+     *
+     * this demo function adds new condition to where clause with sales_stage is NOT IN the list of $this->filterSalesStages
```
Comment (lschaefer-sugarcrm) [inline custom_visibility_7_10/src/custom/data/visibility/FilterOpportunities.php]: I think "with" should be "when"

Comment (lschaefer-sugarcrm) [reply custom_visibility_7_10/src/custom/data/visibility/FilterOpportunities.php]: It looks like you copied a lot of the comments from custom_visibility into this one.  Can you apply the suggestions I had above to this code as well?

File: custom_visibility_7_9/src/custom/data/visibility/FilterOpportunities.php [hunk 17]
```diff
@@ -65,20 +65,30 @@ public function __construct(SugarBean $bean, $params = null)
     }
 
     /**
-     * {@inheritdoc}
+     * orvewrite parent method, SugarVisibility::addVisibilityWhere
```
Comment (lschaefer-sugarcrm) [inline custom_visibility_7_9/src/custom/data/visibility/FilterOpportunities.php]: Same for 7.9.  Can you apply suggestions from custom_visibility to this section of the code as well?

**Review by lschaefer-sugarcrm — CHANGES_REQUESTED** [review]
Thanks for adding the doc!  This will be really helpful!

File: custom_visibility_7_10/src/custom/data/visibility/FilterOpportunities.php [hunk 18]
```diff
@@ -168,23 +187,47 @@ protected function getTableAlias()
...
         // no special analyzers needed
     }
 
     /**
-     * {@inheritdoc}
+     * Implementation of ElasticStrategyInterface interface
+     *
+     * Build Elasticsearch mapping, which defines how a document's fields are stored and indexed in Elasitc engine
+     *
+     * This demo function adds a new not-analyzed field, 'visibility_sales_stage', to Elasitc engine
```
Comment (lschaefer-sugarcrm) [inline custom_visibility_7_10/src/custom/data/visibility/FilterOpportunities.php]: Typo on 208 and 210: should be Elastic

File: custom_visibility_7_9/src/custom/data/visibility/FilterOpportunities.php [hunk 19]
```diff
@@ -203,7 +254,15 @@ public function elasticGetBeanIndexFields($module, Visibility $provider)
...
-     * {@inheritdoc}
+     * Implementation of ElasticStrategyInterface interface
+     *
+     * Add visibility filters
+     *
+     * This demo function adds elastic filter with condition 'filter_sales_stages' NOT IN $this->filterSalesStages
+     *
+     * @param User $user, the user who requests data
+     * @param \Elastica\Filter\Bool $filter, elastic BoolFilter filter
+     * @param VSugarcrm\Sugarcrm\Elasticsearch\Provider\Visibility\Visibility $provider, Elasticsearch Visibility Provider
```
Comment (lschaefer-sugarcrm) [inline custom_visibility_7_9/src/custom/data/visibility/FilterOpportunities.php]: Should the "V" be included as part of "VSugarcrm"?

**Review by lschaefer-sugarcrm — APPROVED** [review]
Thanks for working on this, JZ!

File: custom_visibility/src/custom/data/visibility/FilterOpportunities.php (hunk 6 above)
Comment (mmarum-sugarcrm) [reply custom_visibility/src/custom/data/visibility/FilterOpportunities.php]: I guess a better question is when would this be likely?

Comment (jzhu-sugar) [reply custom_visibility/src/custom/data/visibility/FilterOpportunities.php]: if $where is empty, there is nothing to add to existing WHERE clause
